bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -330,54 +333,216 @@ private int readFromLogChannel(long entryLogId, BufferedReadChannel channel, Byt
+                    .build(readChannelLoader);
+        }
+    };
+
+    @VisibleForTesting
+    long getReadChannelCacheExpireTimeMs() {
+        return readChannelCacheExpireTimeMs;
+    }
+
+    @VisibleForTesting
+    CacheLoader<Long, BufferedReadChannel> getReadChannelLoader() {
+        return readChannelLoader;
+    }
+
+    private final  CacheLoader<Long, BufferedReadChannel> readChannelLoader =
+            new CacheLoader<Long, BufferedReadChannel> () {
+        public BufferedReadChannel load(Long entryLogId) throws Exception {
+
+            return getChannelForLogId(entryLogId);
+
         }
     };
 
     /**
-     * Each thread local buffered read channel can share the same file handle because reads are not relative
-     * and don't cause a change in the channel's position. We use this map to store the file channels. Each
-     * file channel is mapped to a log id which represents an open log file.
+     * FileChannelBackingCache used to cache RefCntFileChannels for read.
+     * In order to avoid get released file, adopt design of FileInfoBackingCache.
+     * @see FileInfoBackingCache
      */
-    private final ConcurrentMap<Long, FileChannel> logid2FileChannel = new ConcurrentHashMap<Long, FileChannel>();
+    class FileChannelBackingCache {
+        static final int DEAD_REF = -0xdead;
+
+        final ConcurrentHashMap<Long, CachedFileChannel> fileChannels = new ConcurrentHashMap<>();
+
+        CachedFileChannel loadFileChannel(long logId) throws IOException {
+            CachedFileChannel cachedFileChannel = fileChannels.get(logId);
+            if (cachedFileChannel != null) {
+                boolean retained = cachedFileChannel.tryRetain();
+                assert(retained);
+                return cachedFileChannel;
+            }
+            File file = findFile(logId);
+            // get channel is used to open an existing entry log file
+            // it would be better to open using read mode
+            FileChannel newFc = new RandomAccessFile(file, "r").getChannel();
+            cachedFileChannel = new CachedFileChannel(logId, newFc);
+            fileChannels.put(logId, cachedFileChannel);
+            boolean retained = cachedFileChannel.tryRetain();
+            assert(retained);
+            return cachedFileChannel;
+        }
+
+        /**
+         * close FileChannel and remove from cache when possible.
+         * @param logId
+         * @param fc
+         */
+        private void releaseFileChannel(long logId, CachedFileChannel fc) {
+            if (fc.markDead()) {
+                try {
+                    fc.fc.close();
+                } catch (IOException e) {
+                    LOG.warn("Exception occurred in ReferenceCountedFileChannel"
+                            + " while closing channel for log file: {}", fc);
+                } finally {
+                    IOUtils.close(LOG, fc.fc);
+                }
+                fileChannels.remove(logId);
```

Comment:
No guarantee that the CachedFileChannel being removed from fileChannels is actually the one referenced by fc.

Reply:
I'll use lock to avoid multi write/delete operations to the fileChannels.

Reply:
it shouldn't need locking. there's a remove method that takes the key and the value, and only removes if both match.

Reply:
okay, then I wonder why you still use `writeLock` in `releaseFileInfo` method? see [code](https://github.com/apache/bookkeeper/blob/26c3379c9fdd812ef36da6176d83e26179705535/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/FileInfoBackingCache.java#L80-L90). Can we reduce lock in there? or the lock is a must when change the state of concurrentHashMap?

Reply:
The write lock is needed because marking a fileinfo as dead, flushing its contents to disk and removing it from the map needs to be an atomic operation. If the lock wasn't there, then we could mark as dead, then before removing from map another thread could call loadFileInfo() and end up returning a dead fileinfo (tryRetain would fail). this would put the caller into a tight loop. We could, i think, remove the lock entirely from this class, but I'm very reluctant to mess with it, since this stuff is hard to get right (and it's often very hard to tell that you got it wrong).